src/eos-updater-apply.c

```
@@ -73,41 +77,592 @@ get_test_osname (void)
+        }
+    }
+
+  return g_steal_pointer (&resolved_flatpaks);
+}
+
+/* Cleanup any flatpaks in flatpaks_to_deploy up until right_bound. This
+ * function is called in cases where one of the flatpaks failed to pull
+ * for some reason and we need to roll back the other pulled flatpaks. This
+ * function "never fails", in the sense that if a rollback fails
+ * the best we can do is just warn about it and move on. */
+static void
+cleanup_undeployed_flatpaks_in_installation (FlatpakInstallation *installation,
+                                             GPtrArray           *flatpaks_to_deploy,
+                                             gsize               right_bound)
+{
+  gsize i = 0;
+
+  for (; i < right_bound; ++i)
+    {
+      FlatpakRemoteRef *to_deploy = g_ptr_array_index (flatpaks_to_deploy, i);
+      const char *remote_name = flatpak_remote_ref_get_remote_name (to_deploy);
+      g_autofree gchar *ref = flatpak_ref_format_ref (FLATPAK_REF (to_deploy));
+      g_autoptr(GError) error = NULL;
+
+      if (!flatpak_installation_remove_local_ref_sync (installation,
+                                                       remote_name,
+                                                       ref,
+                                                       NULL,
+                                                       &error))
+        g_warning ("Couldn't clean up undeployed ref %s:%s: %s",
+                   remote_name,
+                   ref,
+                   error->message);
+    }
+}
+
+static void
+cleanup_undeployed_flatpaks (GPtrArray *flatpaks_to_deploy,
+                             gsize     right_bound)
+{
+  g_autoptr(GError) cleanup_error = NULL;
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (NULL,
+                                                                          &cleanup_error);
+
+  if (!installation)
+    g_warning ("Unable to clean up undeployed flatpaks, couldn't get installation: %s",
```

Why not propagate this `GError` upwards instead of warning here?

---

This is already an error handling path - if we encounter more errors I think the policy here should be to just complain rather than throwing again.

---

Hmm okay. I think it makes sense to make this function propagate, but we can't make `cleanup_undeployed_flatpaks_in_installation` propagate since one of its failure points happens inside a loop (and we wouldn't want to block cleaning up other undeployed flatpaks in case the first failed).